src/Microsoft.SqlTools.ServiceLayer/Connection/ListDatabaseRequestHandler.cs

```diff
@@ -4,192 +4,177 @@
-            databaseInfo.Options[ListDatabasesRequestDatabaseProperties.SizeInMB] = database.Size.ToString();
-        }
-    }
-
-    /// <summary>
-    /// Standalone SQL Server database detail handler
-    /// </summary>
-    class SqlServerDatabaseDetailHandler : BaseDatabaseDetailHandler
-    {
-        protected override void SetProperties(SMO.Database database, DatabaseInfo databaseInfo)
-        {
-            base.SetProperties(database, databaseInfo);
-            databaseInfo.Options[ListDatabasesRequestDatabaseProperties.LastBackup] = (database.LastBackupDate == default(DateTime)) ? "" : database.LastBackupDate.ToString("yyyy-MM-dd hh:mm:ss");
-        }
-    }
-
-    /// <summary>
-    /// SQL DB database detail handler
-    /// </summary>
-    class SqlDBDatabaseDetailHandler : BaseDatabaseDetailHandler
-    {
-
-    }
+	public class ListDatabasesRequestDatabaseProperties
+	{
+		public const string Name = "name";
+		public const string SizeInMB = "sizeInMB";
+		public const string State = "state";
+		public const string LastBackup = "lastBackup";
+	}
+
+	/// <summary>
+	/// Factory class for ListDatabasesRequest handler
+	/// </summary>
+	static class ListDatabaseRequestHandlerFactory
+	{
+		public static IListDatabaseRequestHandler getHandler(bool includeDetails, bool isSqlDB)
+		{
+			if (!includeDetails)
+			{
+				return new DatabaseNamesHandler();
+			}
+			else if (isSqlDB)
+			{
+				return new SqlDBDatabaseDetailHandler();
+			}
+			else
+			{
+				return new SqlServerDatabaseDetailHandler();
+			}
+		}
+	}
+
+	/// <summary>
+	/// Interface of ListDatabasesRequest handler
+	/// </summary>
+	interface IListDatabaseRequestHandler
+	{
+		ListDatabasesResponse HandleRequest(ISqlConnectionFactory connectionFactory, ConnectionInfo connectionInfo);
+	}
+
+	/// <summary>
+	/// Base handler
+	/// </summary>
+	abstract class ListDatabaseRequestHandler<T> : IListDatabaseRequestHandler
+	{
+		private static readonly string[] SystemDatabases = new string[] { "master", "model", "msdb", "tempdb" };
+
+		public ListDatabasesResponse HandleRequest(ISqlConnectionFactory connectionFactory, ConnectionInfo connectionInfo)
+		{
+			ServerConnection serverConnection = ConnectionService.OpenServerConnection(connectionInfo);
+			SMO.Server server = new SMO.Server(serverConnection);
+
+			var results = new List<T>();
+			ListDatabasesResponse response = new ListDatabasesResponse();
+
+			// Using this function to set the default init properties of the server so that those things can be fetched in a single query.
+			setDefaultServerInitProperties(server);
+
+			foreach (SMO.Database db in server.Databases)
+			{
+				results.Add(CreateItem(db));
+			}
+
+			results = results.Where(s => SystemDatabases.Any(x => this.NameMatches(x, s))).Concat(
+						 results.Where(s => SystemDatabases.All(x => !this.NameMatches(x, s)))).ToList();
+
+			SetResponse(response, results.ToArray());
+			serverConnection.Disconnect();
```

Comment:
should this be in a finally block?

Reply:
previously i left a comment about putting the serverConection into a using statement.

Reply:
Unfortunately, I cannot use the using statement for serverConnection. I'll add that in a finally block.